receiver/dcgmreceiver/client.go

```diff
@@ -372,83 +361,79 @@ func (client *dcgmClient) cleanup() {
 	client.logger.Info("Shutdown DCGM")
 }
 
-func (client *dcgmClient) getDeviceModelName(gpuIndex uint) string {
-	return client.devicesModelName[gpuIndex]
-}
-
-func (client *dcgmClient) getDeviceUUID(gpuIndex uint) string {
-	return client.devicesUUID[gpuIndex]
-}
-
-func (client *dcgmClient) collectDeviceMetrics() (map[uint][]dcgmMetric, error) {
-	var err scrapererror.ScrapeErrors
-	gpuMetrics := make(map[uint][]dcgmMetric)
-	for _, gpuIndex := range client.deviceIndices {
-		client.logger.Debugf("Polling DCGM daemon for GPU %d", gpuIndex)
-		retry := true
-		for i := 0; retry && i < client.maxRetries; i++ {
-			fieldValues, pollErr := dcgmGetLatestValuesForFields(gpuIndex, client.enabledFieldIDs)
-			client.logger.Debugf("Got %d field values", len(fieldValues))
-			if pollErr == nil {
-				gpuMetrics[gpuIndex], retry = client.appendMetrics(gpuMetrics[gpuIndex], gpuIndex, fieldValues)
-				if retry {
-					client.logger.Warnf("Retrying poll of DCGM daemon for GPU %d; attempt %d", gpuIndex, i+1)
-					time.Sleep(client.pollingInterval)
-					continue
-				}
-				client.logger.Debugf("Successful poll of DCGM daemon for GPU %d", gpuIndex)
-			} else {
-				msg := fmt.Sprintf("Unable to poll DCGM daemon for GPU %d on %s", gpuIndex, pollErr)
-				client.issueWarningForFailedQueryUptoThreshold(gpuIndex, "all-profiling-metrics", msg)
-				err.AddPartial(1, fmt.Errorf("%s", msg))
-			}
-		}
+// collect will poll dcgm for any new metrics, updating client.devices as appropriate
+// It returns the estimated polling interval.
+func (client *dcgmClient) collect() (time.Duration, error) {
+	client.logger.Debugf("Polling DCGM daemon for field values")
+	fieldValues, pollTime, err := dcgmGetValuesSince(dcgm.GroupAllGPUs(), client.enabledFieldGroup, client.lastSuccessfulPoll)
```

Comment:
My thought was that if dcgm gives us data for a new GPU we'll handle it correctly, so it's harmless whether we get data for other GPUs or not.

Reply:
I'm more worried about errors when trying to query data from unsupported GPUs (I mean, they must have been excluded from the list of supported devices for a reason, right?).

Reply:
We already have to worry about that if there are two GPUs with different sets of supported metrics, so assuming that works, I think it should also be fine if the new GPU has fewer supported metrics.

Reply:
That's not what I was saying. Suppose one of the devices is completely unsupported. Would attempting to query that device trigger a library error, rather than a sequence of "unsupported" values?

Reply:
That's certainly my assumption. At the top level API, there's separate `dcgmGetAllDevices` and `dcgmGetAllSupportedDevices` functions. But there's only one enum, `DCGM_GROUP_ALL_GPUS`. The docs for `dcgmGetValuesSince_v2` just say

> Alternatively, pass in the group id as `DCGM_GROUP_ALL_GPUS` to perform operation on all the GPUs or `DCGM_GROUP_ALL_NVSWITCHES` to perform the operation on all NvSwitches.

If that fails if there's also an unsupported GPU in the system, that seems like it would be unusable. Also, `DCGM_ST_NOT_SUPPORTED` is not documented as a possible return value for `dcgmGetValuesSince_v2` (which it is for other functions), so I think that also indicates it's okay to call with unsupported GPUs present.

If you want a reliable answer, we'll have to ask NVIDIA.

Reply:
That device group is only used in the constructor and isn't available in the rest of the client. I guess you're asking me to store it in the client object for use later.

I was intending to make this the first step towards removing the device enumeration in the constructor entirely, but it's not worth adding even more blockers to the review.

Reply:
Yeah, I'd rather proceed slowly on this one and confirm with NVIDIA first. Thanks!